### ISSUE TITLE
Evaluate necessity of human check circumvention functionality in the crawler

### ISSUE THREAD

**Issue by SebastianZimmeck** [privacy-tech-lab/gpc-web-crawler#135]
Before the next crawl (#118), we should look into what caused the divergence of crawl results from the manual analysis results in our [most recent analysis](https://docs.google.com/spreadsheets/d/1jRp9exj2Sc5ShaMijdLObWQ0Q-tJ-lck?rtpof=true&usp=drive_fs). See the red-labeled fields:

<img width="431" alt="Screenshot 2024-09-18 at 9 03 57 AM" src="https://github.com/user-attachments/assets/afc81689-91ab-463a-b388-a923c537388c">

Do increased timeouts help? Maybe, some sites were not fully loaded before the data was captured. Are there other parameters that we can fine-tune to improve accuracy?

@franciscawijaya will take the lead here and work with @natelevinson10 and @eakubilo before starting the next crawl.

**Comment by franciscawijaya** [comment]
Some of my findings: 
1) Optanonconsent after GPC
- increased timeout did not change the crawl data (especially for Optanonconsent **after** gpc) because there is a human check error
- I noticed that the websites that had discrepancies with the manual check for Optanonconsent_after_gpc had the same human check error power by PerimeterX (see attached below) 
- I tried to bypass the human check error by pressing and holding during the crawl but was not able to bypass it; my conclusion is that this particular human check mechanism is that it recognized that we are accessing the site using a selenium crawler (as it is mentioned that some of the reasons of the limited access to the page was because either javascript is disabled or the browser does not support cookies -- suggesting that the human check recognized that there is a possibility of automation in the browser) 
- I also tried to inspect storage while the crawl is ongoing with the human check and realize that in the storage, the value of the cookie is also zero. Hence, we can conclude that the human check page does not reflect the cookie value of the actual site

2) Pridecounseling.com redirect
- the crawl did not capture the data of betterhelp.com (the site that pride counseling.com is being redirected to when done manually)
- It instead logged it in the errorloggkng as an InsecureCertificateError
- While during manual check, we are redirected to betterhelp.com, the crawler instead only access a blank page when trying to access pride counseling.com 

The good news are that
- all the human check errors that caused the optanonconsent after gpc to still be 0 are correctly logged as humancheckerrors  in error logging. 

Given that this error is correctly logged in the error logging and would only be able to be accessed manually, I think it would still be okay to start the crawl. Should we encounter this in our crawl, this has better informed us about the possible reason of cases where optanonconsent cookies are 0 both before and after GPC for sites that are also flagged with human check errors. I will begin the California crawl shortly. 

<img width="1440" alt="Screenshot 2024-09-23 at 3 27 57 AM" src="https://github.com/user-attachments/assets/26e61d0f-303e-4df3-b2b8-0833fb0479a6">

**Comment by SebastianZimmeck** [comment]
@franciscawijaya, for the Optanonconsent after GPC:

1. Were some of the sites also in our previous crawls? If so, what were the results then?
2. Do you get the same results for Colorado, California, another VPN? In other words, is this VPN-dependent?

@natelevinson10 and @eakubilo, if you have a chance, please also look into this issue? This is the most important issue at the moment. It would be good to get a good understanding before we start the next crawl.

**Comment by SebastianZimmeck** [comment]
@natelevinson10 is [saying](https://github.com/privacy-tech-lab/gpc-web-crawler/issues/136#issuecomment-2375468893):

> Did a quick review of the manual data we collected a couple of weeks ago and targeted instances of a mismatch (marked in red) signaling our ground truth was different than the crawl data. I used several VPN locations (California, multiple Colorado, Virginia, and no VPN (CT)) and gave ample time to let all of the site content to load.
> 
> I was not able to find a single instance of our manual data changing from what we had reported, except for bumble.com 's USPapi_before being "1YNN" instead of the reported "1YYN", and I would chalk this up to a manual error on our end. It would seem that for the mismatches of data from crawl to manual, the manual data is more accurate.

**Comment by franciscawijaya** [comment]
> 1. Were some of the sites also in our previous crawls? If so, what were the results then?

I wanted to be precise and check the progression of the data across the different rounds of Crawl that we did and these are the findings: 
**Dickies.com**: Dec-April (isGpcEnabled=1) [matches Manual Data], June Crawl (isGpcEnabled=0) [matches recent Crawl Data] --> **high priority**: sudden change in the Crawl output 
**Altrarunning.com**: Dec-June (isGpcEnabled=1) [matches Manual Data] --> **high priority**: the Crawl output has been accurate since the first crawl so it is strange that we are getting isGpcEnabled = 0 now 
**Smartwool.com**: Dec-June (isGpcEnabled=1) [matches Manual Data] --> **high priority**: the Crawl output has been accurate since the first crawl so it is strange that we are getting isGpcEnabled = 0 now 
**Goodrx.com:** NULL for uspapiaftergpc [matches recent Crawl Data] --> low priority: the difference between Crawl & Manual has been present since the first Crawl 
**Redrobin.com**: Dec-Feb (isGpcEnabled=0) [matches Recent Crawl Data], April-June (isGpcEnabled=1) [matches Manuall Data] --> **high priority**: it's strange that the Crawl has gotten an outputof isGpcEnabled = 1 after previously ouputting 0 but is now back to 0

Analysis: 
While there are only 5 overlapping sites, their outputs in the past crawls have mostly been the correct ones, with the exception of Dickies.com so it's pretty strange that we got the opposite for our recent small crawl. 

> 2\. Do you get the same results for Colorado, California, another VPN? In other words, is this VPN-dependent?

Since I used Colorado VPN for the recent small crawl, I also have done a crawl focusing on these focus sites with California crawl. However, after analyzing the result with California VPN, I realized that the VPN could be a potential problem as it gave minor differences in the outputs. Hence what I will do is redo these focus mini crawls with more VPN IP addresses to have more data to compare (ie. using more than 1 Colorado and California VPN) and confirm before writing the analysis here.

**Comment by SebastianZimmeck** [comment]
Thanks, @franciscawijaya!

> While there are only 5 overlapping sites, their outputs in the past crawls have mostly been the correct ones, with the exception of Dickies.com so it's pretty strange that we got the opposite for our recent small crawl.

That is helpful to know! So, before starting the next crawl we should try to understand what the reasons for these performance drops are.

> I realized that the VPN could be a potential problem as it gave minor differences in the outputs. Hence what I will do is redo these focus mini crawls with more VPN IP addresses to have more data to compare 

Yes, that is a good point to try.

@natelevinson10, can you coordinate with @franciscawijaya and also look into this as a team?

**Comment by franciscawijaya** [comment]
As mentioned during our call, I redid the crawl using 6 different VPN from Colorado and California and the manual-crawl data discrepancy is unfortunately still present. Hence, I concluded to rule out a VPN issue. 

I also looked through the particular human check error from all these different sites as all of their human checks were powered by the same company, [PerimeterX ](https://www.humansecurity.com/human-perimeterx-merger?utm_source=google&utm_medium=cpc&utm_campaign=brand&utm_term=perimeterx&utm_content=615179403772&gad_source=1&gclid=Cj0KCQjwmOm3BhC8ARIsAOSbapUlEElsMiUBsrmScAXACLz_uffdxQGRk_r4qhJFyzVkGxm9sVGunlUaAiUOEALw_wcB). Examining the past crawl results, I started to wonder if the cause of this discrepancy is a recent integration of this particular human check. I have also researched and browsed through the possibility of bypassing this human check to find [useful resources](https://scrapfly.io/blog/how-to-bypass-perimeterx-human-anti-scraping/).

In the next couple of weeks, @natelevinson10 and I will try to examine the possibility of bypassing the human check error. As it stand right now, our crawler is correctly identifying and flagging the human check error so we will now be exploring the possibility of bypassing human checks with our crawler.

**Comment by SebastianZimmeck** [comment]
Nice work, @franciscawijaya! 

Our approach is two-pronged:
1. Check if we can add human check circumvention functionality to our crawler (at the moment, we are only logging human checks; we are not trying to circumvent those; this is not only true for PerimiterX but for any human check service)
2. If we come to the conclusion (after, say, two weeks) that it is too difficult to circumvent human checks, make sure that the logging of human checks is accurate and captures PerimeterX so that we can exclude data from sites with human checks post-crawl in the data analysis

Since we have no human check circumvention functionality at all in the crawler, this issue is not limited to PerimeterX but also other human check services. 

As a starting point for a potential human check functionality @natelevinson10 will do a mini crawl of sites with human check known to us and observe their behavior, which can then point towards circumvention strategies.

@natelevinson10 and @franciscawijaya will coordinate who does what on this issue.

**Comment by natelevinson10** [comment]
I compiled a list of 100 sites that were tagged as "HumanCheckError" from our June 2024 crawls batches 1-2. This list is comprised of 97 sites total. I ran an initial crawl with timeouts set to 10,000 (for crawler) and 5,000 (for analysis), and have uploaded the results in a new "Human_Check_Analysis" folder in the drive.

It seems like the majority (~80%) of the sites threw a human verification error. I think it could be a good idea to crawl the same sites on a few different VPN locations to see if the errors are consistent across numerous IP addresses. This is something I could look into next week

**Comment by SebastianZimmeck** [comment]
> I compiled a list of sites that were tagged as "HumanCheckError" from our June 2024 crawls batches 1-2. This list is comprised of 97 sites total. I ran an initial crawl with timeouts set to 10,000 (for crawler) and 5,000 (for analysis), and have uploaded the results in a new "Human_Check_Analysis" folder in the drive.

@natelevinson10, what is the path to this folder? Can you also link it here? I see a "Human_Verification_Crawl" folder. Is that the folder you mean? 

> It seems like the majority (~80%) of the sites threw a human verification error. 

So, these were sites that all had a human check in June 2024 and 20% of those no longer have such?

> I think it could be a good idea to crawl the same sites on a few different VPN locations to see if the errors are consistent across numerous IP addresses. This is something I could look into next week

Let's hold off of that for the time being. It is likely that a different VPN IP addresses will be also blocked.

Just to remind what the point of this whole exercise is: per our [results](https://docs.google.com/spreadsheets/d/1jRp9exj2Sc5ShaMijdLObWQ0Q-tJ-lck/edit?usp=sharing&ouid=112157414060543752223&rtpof=true&sd=true) we see the following (apparently due to a human check; is the human check the issue for these discrepancies?):

<img width="375" alt="Screenshot 2024-10-06 at 1 06 54 PM" src="https://github.com/user-attachments/assets/b5e48984-3e49-45de-9d58-9fb25ad9498d">

Are these occurrences of new human checks that we have not seen before so frequent that we need to address those by trying to add circumenvention functionality to the crawler?

**Comment by natelevinson10** [comment]
Here is the path to the folder: 
https://drive.google.com/drive/u/1/folders/1wdAvF_-Tkfux-lD0rDahZzZ3Az_NQdbM

80% of the sites had a human check error, and a hanfdul errror'd out, meaning ~15 sites did not have a human check error despite having one earlier. As for the CPA GPC data that we manually reviewed, it would be a good idea to run crawls on just these sites and see if they every throw a human verification. Possibly crawling the few sites with data mismatches several times, changing the IP each time and compiling the results. This is definitely something I can do over the next few days.

**Comment by SebastianZimmeck** [comment]
> As for the CPA GPC data that we manually reviewed, it would be a good idea to run crawls on just these sites and see if they every throw a human verification. ... This is definitely something I can do over the next few days.

Excellent, that is a good idea! Let's do that.

**Comment by natelevinson10** [comment]
I ran a crawl on the 9 sites showing data mismatches on our CO crawl vs. manual data sheet (here is the link to the crawl in the drive: [here](https://drive.google.com/drive/u/1/folders/1YTeYbB6XCQy7ghPDldwo5y6ooEsa3mpt) ) After looking at the results, it is pretty clear that human check errors are causing these data discrepencies.

For this crawl, I ran the with my local IP (batch 1), a CO vpn (batch 2), and CA vpn (batch 3), all on the same 9 sites. After looking at the results, here are is what I can conclude. First, for bumble.com, I previously mentioned the data mismatch was likely a manual error when we collected the ground truth data. This was confirmed when all 3 batches had the same data that matched our CO crawl data, showing that our manually collected uspapi_before entry is wrong. 

For rest of the sites, dickies, altarunning, smartwool, goodrx, and redrobin all were flagged as having a human check. Also, coorstek had an insecure cerficate error. The data discrepencies between crawl and manual data is certainly a result of human check errors in this case. The other two sites, crocs and coloradosun, flagged no errors, but I noticed all of the key/value pairs of data for these sites are null across all 3 crawls. This makes me question if there was a human check error that ended up going unflagged, which is something I plan to look into. It should be worthwhile to do another crawl and pay attention to the chromium browser to see if any human validation errors show up.

**Comment by SebastianZimmeck** [comment]
> For rest of the sites, dickies, altarunning, smartwool, goodrx, and redrobin all were flagged as having a human check. Also, coorstek had an insecure cerficate error. 

So, there was no difference for these sites between your conditions 1, 2, and 3?

> The data discrepencies between crawl and manual data is certainly a result of human check errors in this case. 

Did the sites display a human check, log a human check, or what is the reason why you think that?

**Comment by natelevinson10** [comment]
There is slight variance between the 3 runs on those sites, namely that my local IP showed gpc codes for dickies, and a uspapi_after code for redrobin, while the other 2 didn't. Also, goodrx got flagged for a human check on runs 1 and 3 but not 2, and redrobin for flagged for a human check on run 2 but not 1 and 3. I would assume both of these results are because using different IPS plays a large role on when / how quickly a human check is prompted. And I would also assume a VPNs IP address will be much easier to flag than that of a real local device.

**Comment by SebastianZimmeck** [comment]
And on this?

>   The data discrepencies between crawl and manual data is certainly a result of human check errors in this case.

Did the sites display a human check, log a human check, or what is the reason why you think that?

**Comment by SebastianZimmeck** [comment]
How does a human check manifests itself? Do you see a screen with a human check, for example?

**Comment by SebastianZimmeck** [comment]
Thanks, @natelevinson10!

We would need to use the California and Colorado VPN connections because that is where the laws apply that we want to target. Within those connections we can still connect to different IP addresses. For example, there are different IP addresses for Denver on Mullvad. So, can you pick, say, three different Denver addresses and check whether your results still hold? 

You can also do it the other way around. For a site that has a human check on a Denver VPN address, can you do a crawl of that one site on each of the others Denver VPN connections and check if one of those do not fail due to the human check?

**Comment by natelevinson10** [comment]
Yes, this sounds like a good idea. As mentioned in the teams channel I am out of the country this weekend so I won't be able to run the crawls before our next meeting on Monday. It is definitely something I can do once I am back though

**Comment by franciscawijaya** [comment]
I have finished running the crawl on one big batch of sites and analyzed the [result.](https://drive.google.com/drive/folders/1VR1Vcc2QX-YkJWeW6HMXjhSscAgpsET8?usp=share_link) Based on the number of the sites flagged as humancheckerror, I have come to realize that humancheckerror is not a significant issue in the grand scheme of things. By comparing the result of a crawl of one batch and comparing it to the result of a crawl of the exact same batch in June, the number of sites being flagged as humancheckerror is around the same(99-104 sites) for a crawl of 1499 sites. Since there is no significant increase in the number of sites flagged with HumanCheckError in a pool of 1499 sites for one big batch of crawl, it would be safe to [conclude](https://docs.google.com/document/d/1ApMrgCTpfCofUtW6AwK5fAj_UkkeRgaZLkC5OGTm9pI/edit?usp=sharing) that our crawl is still functional in the grand scheme of things and figuring out a way to circumvent human checks is not necessary.

**Comment by SebastianZimmeck** [comment]
> I have come to realize that humancheckerror is not a significant issue in the grand scheme of things. 

Excellent to hear!

As I said [above](https://github.com/privacy-tech-lab/gpc-web-crawler/issues/135#issuecomment-2411581772):

> Ideally, this would also involve some spot checks of crawl analysis results of sites that were not flagged as human checks in the crawl. 

So, @franciscawijaya can you manually check some randomly selected sites that did *not* return a human check error? Because the sites in the Colorado crawl for which we saw the `isGpcEnabled=1` discrepancies between manual analysis and crawl were not flagged as human check (or were they?). So, in your crawl any such discrepancies would go undetected. We can discuss tomorrow.

**Comment by SebastianZimmeck** [comment]
@franciscawijaya clarified that the [discrepancies we observed between crawl results and manual ground truth for Colorado](https://github.com/privacy-tech-lab/gpc-web-crawler/issues/135#issuecomment-2395512343) are included in counting of human check errors in the [results above](https://github.com/privacy-tech-lab/gpc-web-crawler/issues/135#issuecomment-2426663511). This percentage of sites with human checks is tolerable, and we do not need to do any [further check](https://github.com/privacy-tech-lab/gpc-web-crawler/issues/135#issuecomment-2427085131). 

The only remaining open point, which @natelevinson10 will examine, is whether we want to try to decrease human checks a bit by selecting California and Colorado IP addresses that are not as often detected as VPN addresses.

**Comment by natelevinson10** [comment]
I have concluded my 10x crawls again, this time running 5 on CA IPs and 5 on CO IPs. Here are the [results](https://docs.google.com/document/d/1CGyWcf_8zRGBdclvpqlX_lHVBOqmx7ugNNDJqEI7ivM/edit): 

Here, the differences between CA and CO are pretty negligible - only zillow, legacy, and realtor had mismatched results, all by a delta of just one. California generated 45 human verification flags while Colorado generated 44. These results definitely indicate that there is not a significant difference between locations, speficially between CA and CO. Tying this in with the findings from my crawl of 10 random IPs, I believe the ultimate conclusion is this: Certain websites are more strict / more likely to throw a human check error, and the IP does not influence whether or not one is presented. After comparing all of the websites between the 20 total crawls, the sites that had human checks on 10/10 in my first crawl also had 10/10 in the second crawl. Additionally, the sites with 6-8 flags of the 10 in the first crawl also had 6-8/10 in the second. This has led me to the conclusion that the likelihood of triggering a human verification check is more dependent on the website’s own security thresholds than the specific IP.

**Comment by SebastianZimmeck** [comment]
Based on the analysis in this issue and @franciscawijaya's view the most likely explanation for the [inconsistent results](https://github.com/privacy-tech-lab/gpc-web-crawler/issues/135#issue-2533695202) is that the sites involved adopted new human checks. These new human checks are correctly flagged in our analysis results. 

@natelevinson10 found that [IP address selection has very little impact](https://github.com/privacy-tech-lab/gpc-web-crawler/issues/135#issuecomment-2440302058). So, there are no changes on how we select an IP address.

The path forward is that we will perform the November crawl (#118). After the first batch for California we perform a spotcheck to identify the performance (#136). We will do such spotchecks on a regular basis, say, every other crawl to detect any decrease in performance that may occur over time and that we would not detect otherwise.
